Lip 003 - Easy Track

grstepanov opened the pull request.

grstepanov committed:
Create lip-3 assets folder

grstepanov committed:
Upload EasyTrack schema

grstepanov committed:
Create lip-3.md

grstepanov committed:
Fix formatting

grstepanov committed:
Header update

grstepanov committed:
Merge branch 'lidofinance:develop' into lip-003

grstepanov committed:
Reviewed to match the final tech design

Updated the motion enactment process desc

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+
+## Problem
+
+Lido DAO governance currently relies on Aragon Voting App model. This means DAO approves or rejects proposals via direct governance token voting. Though transparent and reliable, it is not a convenient way to make decisions only affecting small groups of Lido DAO members. Besides, direct token voting doesn't exactly reflect all the decision-making processes within the Lido DAO and is often used o ...
```
It doesn't seem to be correct that the Lido DAO uses Aragon votings to formalize any decisions — for ones not requiring the on-chain tx we do use snapshots.

grstepanov replied:
Not any, but I believe the Lido DAO does **often** formalize existing consensus via Aragon votings. 
That's what happens with the rewards allocations: first the community sentiment gets measured on the research forum and then rubber-stamped with the omnibus. In fact, it's something already decided but still requiring an on-chain confirmation.

kadmil replied:
In that case, let's add something along the lines of "reached in discussion on research forum or snapshot vote", please. We don't just slap things into Aragon votes, are we?

grstepanov replied:
Fixed wording.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+## Problem
+
+Lido DAO governance currently relies on Aragon Voting App model. This means DAO approves or rejects proposals via direct governance token voting. Though transparent and reliable, it is not a convenient way to make decisions only affecting small groups of Lido DAO members. Besides, direct token voting doesn't exactly reflect all the decision-making processes within the Lido DAO and is often used o ...
+There are a few natural sub-governance groups within the DAO, e.g. node operators committee, financial operations team and LEGO committee. Every day they need to take routine actions only related to their field of expertise. The decisions they make hardly ever spark any debate in the community, and votings on such decisions often struggle to attract wider DAO attention and thus, to pass.
```
to get the quorum & be enacted as the on-chain tx

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+## Proposed solution
+
+We propose Easy Track frictionless motions as a solution to this problem. 
+Easy Track motion is a lightweight voting considered to have passed if the minimum objections threshold hasn’t been exceeded. As opposed to traditional Aragon votings, Easy Track motions are cheaper (no need to vote ‘pro’, token holders only have to vote ‘contra’ if they have objections) and easier to manage (no need to ask broad DAO community vote on proposals sparking no debate). 
```
`threshold hasn’t been exceeded` — let's mention that there's a timeframe there?

grstepanov replied:
Done.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+
+## Sub-governance groups
+
+There's a sub-governance group within the Lido DAO for each of the voting cases listed above. Those groups have been formed intentionally or in a more natural way, and at this point creating and maintaining periodical Aragon votings falls within their remits.
```
As mentioned below, we don't have finance ops team / subgroup currently. There's a finance multisig, but it's a hack job and we'll need a proper one set up.

grstepanov replied:
I think the finance multi-sig owners may be considered an informal sub-governance group at this point. Adding this into the LIP to make it more precise.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+
+# Specification
+
+To create a feature as flexible and extendable as possible, we propose implementing Easy Track functionality within several contracts rather than building a single contract for it. 
```
The maximal flexibility in itself wasn't a prio, but having enough of it to accommodate for current and future "operational" transactions being regulary made from Lido DAO side. Having opt-out gov thing flexible and extendable in itself sounds properly scary.

grstepanov replied:
Like I put it in the LIP, flexibility and extendability were the two primary reasons why we preferred splitting the feature into multiple contracts rather than building a Frankenstein master contract instead. Another reason worth mentioning is having the feature supervised, adding it into the proposal.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+
+A group of core smart contracts will implement basic logic and store data related to Easy Track motions, and a series of separate EVMScriptFactory smart contracts will represent all the supported types of Easy Track motions. 
+This architecture aims to support further extension by adding new types of Easy Track motions without replacing the core of the proposed system.
+1. A motion can be started by calling `createMotion()` function on `EasyTrack` core contract. 
```
I'd suggest adding small intro paragraph "that's what regular motion operation looks like" — we're telling the story of overall architecture and then jump to smaller detail right ahead.

grstepanov replied:
Added.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+1. A motion can be started by calling `createMotion()` function on `EasyTrack` core contract. 
+2. Upon motion creation, `EasyTrack` calls `createEvmScript()` function on the corresponding `EVMScriptFactory` contract passing all the data required to generate a motion enactment EVM script.
+3. `EVMScriptFactory` smart contract generates the script and returns it to the `EasyTrack` core contract.
+4. `EasyTrack` contract conducts the motion according to the motion settings (i.e. motion duration, objections threshold and maximum active motions count). While active, the motion can be canceled at any time by the address that has started it.
```
May want to mention explicitly that the motion could be cancelled by the address started it at any time before the execution

grstepanov replied:
Done.

kadmil replied:
still not clear whether pending (timeframe passed, not enough reject-votes to reject, not executed) motion could be cancelled by the creator

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+4. `EasyTrack` contract conducts the motion according to the motion settings (i.e. motion duration, objections threshold and maximum active motions count). While active, the motion can be canceled at any time by the address that has started it.
+5. As soon as the motion duration expires, it becomes possible to enact the motion. Anyone can enact the motion from that moment on.
+6. To enact the motion, the `EasyTrack` contract recreates the enactment EVM script and passes it to the `EVMScriptExecutor` smart contract for execution.
+7. `EVMScriptExecutor` holds all the permissions to run valid enactment EVM scripts passed from the `EasyTrack` smart contract (or Aragon voting app). 
```
Do we need Aragon voting mentioned here? Even if so, a separate point could be a bit better, imo

grstepanov replied:
Removed the Aragon mention, covering it in a new paragraph about ET<>Aragon relations.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+### `EasyTrackStorage` contract
+
+`EasyTrackStorage` contract stores all the variables of Easy Track system. It inherits from battle-tested `Initializable`, `PausableUpgradeable` and `AccessControlUpgradeable` OpenZeppelin contracts.
+On the other hand, `MotionSettings` and `EVMScriptFactories` inherit from `EasyTrackStorage` to make variables stored in the latter accessible to their methods.
```
If those are two different contracts, would we be forced to have two sets of variables, which should be just the same? This sounds a bit unclear & may cause ops hazard.

grstepanov replied:
Actually, this contract has been removed alongside the upgradeability, dropping this paragraph altogether.
The variables are now stored in MotionSettings, EVMScriptFactoriesRegistry and EasyTrack contracts right next to the logic.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+______
+## `EVMScriptFactory` smart contracts
+
+Initially, five `EVMScriptFactory` contracts will represent five Easy Track motion types.
```
We've mentioned three types on the start; I know that we'll have all five, but here it's worth explaining the difference, imo

grstepanov replied:
I don't see a contradiction here, as we mentioned three types of votings we would like to wrap into motions. Wrapping resulted in five motions, which is pretty obvious from the list right below this line.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+## `EVMScriptExecutor` contract
+
+The EVM scripts generated by the script factory contracts can be executed by a separate `EVMScriptExecutor` contract. It implements only `executeEVMScript()` method and accepts scripts built following the Aragon's EVM script structure. Current implementation uses [CallsScript.sol](https://github.com/aragon/aragonOS/blob/next/contracts/evmscript/executors/CallsScript.sol) contract of the Aragon fr ...
+`EVMScriptExecutor` permissions list is the same as the one of Aragon Voting App. While sharing the EVM script structure, both `EasyTrack` contract and Aragon Voting App are able to execute scripts using `EVMScriptExecutor` contract.
```
Would it have the same roles assigned from the get-go? If so, let's make it more explicit?

grstepanov replied:
Specified in the Roles management and connection to the current Aragon setup paragraph

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+The EVM scripts generated by the script factory contracts can be executed by a separate `EVMScriptExecutor` contract. It implements only `executeEVMScript()` method and accepts scripts built following the Aragon's EVM script structure. Current implementation uses [CallsScript.sol](https://github.com/aragon/aragonOS/blob/next/contracts/evmscript/executors/CallsScript.sol) contract of the Aragon fr ...
+`EVMScriptExecutor` permissions list is the same as the one of Aragon Voting App. While sharing the EVM script structure, both `EasyTrack` contract and Aragon Voting App are able to execute scripts using `EVMScriptExecutor` contract.
+___
+# Sanity and security
```
Maybe `security and sanity checks`?

grstepanov replied:
Done.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+___
+# Sanity and security
+
+The proposed Easy Track Motions design involves timelock as a native security feature. Any suspicious motion can be rejected or canceled within a relatively long motion duration period. To prevent malicious actions, we propose putting the following additional limitations in place (numeric values fixed based on community feedback from v1):
```
Do we have a link on the feedback?

grstepanov replied:
Most of this discussion happened on research forum [here](https://research.lido.fi/t/lip-3-easy-track-motions-for-dao-routine-operations/680).
We have a discussion link in the LIP header already, but I am adding it here as well.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,152 @@
+
+- It should be impossible to set objections threshold at more than 5% of the total LDO supply. Nevertheless, the default objections threshold should be set a lot lower at 0.5% of the total LDO supply.
+- It should be impossible to set motion duration at less than 48 hours. For any motion, there should be enough time for DAO to submit objections and reject the motion. 
+- It should be impossible to spam motions. The default limit for simultaneously active motions will be set at 12 and it can only be increased by the DAO to up to 24 motions at a time.
```
Let's add explicitly that the limit is on all types of motions in total, and not for every time in particular?

grstepanov replied:
Good point.

kadmil reviewed (commented):
Added a couple nitpicks, but overall seems good! I'd rephrase the intro part & focus on safety measures & rules we do have around 1) running current types of easy tracks; 2) changing available types of Easy Tracks (btw, how are new entities allowed to creating Easy Tracks?); 3) roles management & connection to Aragon setup we do have.

grstepanov committed:
Update wording based on first LIP review (part 1)

grstepanov committed:
Update LIP-3 based on review (part 2)

- Added paragraph on adding new Easy Track types
- Added paragraph on roles management and connection to current Aragon setup

grstepanov committed:
Update LIP-3 based on review (part 3)

- Dropped EasytrackStorage contract (removed alongside the upgradeability)
- Minor typo fix

grstepanov committed:
Delete outdated schema

grstepanov committed:
Upload new easytrack schema

grstepanov committed:
Fix the list of core contracts, replace the schema

grstepanov committed:
Clean-up LIP to match the final spec

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,160 @@
+
+## Sub-governance groups
+
+There's a formal or an informal sub-governance group within the Lido DAO for each of the voting cases listed above. Those groups have been formed intentionally or in a more natural way, and at this point creating and maintaining periodical Aragon votings falls within their remits.
```
For the last part, it's not creating the votings, but ensuring the necessary transactions get processed through means of Aragon vote. The groups don't start the votes in general, and only manage actions required for their respective ops

grstepanov replied:
Fixed.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,160 @@
+2. Upon motion creation, `EasyTrack` calls `createEvmScript()` function on the corresponding `EVMScriptFactory` contract passing all the data required to generate a motion enactment EVM script.
+3. `EVMScriptFactory` smart contract generates the script and returns it to the `EasyTrack` core contract.
+4. `EasyTrack` contract conducts the motion according to the motion settings (i.e. motion duration, objections threshold and maximum active motions count). While active, the motion can be canceled at any time by the address that has started it.
+5. As soon as the motion duration expires, it becomes possible to enact the motion. Anyone can enact the motion from that moment on.
```
Need a point on what happens if objection threshold is reached

grstepanov replied:
Added the point.

kadmil commented on LIPS/lip-3.md:
```diff
@@ -0,0 +1,160 @@
+- It should be impossible to set objections threshold at more than 5% of the total LDO supply. Nevertheless, the default objections threshold should be set a lot lower at 0.5% of the total LDO supply.
+- It should be impossible to set motion duration at less than 48 hours. For any motion, there should be enough time for DAO to submit objections and reject the motion. 
+- It should be impossible to spam motions. The default limit for simultaneously active motions will be set at 12 and it can only be increased by the DAO to up to 24 motions at a time. This value refers to total number of active motions of all types.
+- Pause lever should be added. It is proposed to set up an Easy Track security multi-sig to pause Easy Track in case of emergency.
```
May want to specify what the pause entails: no new motions, active & pending can't be executed, ones passing during the pause can't be executed after unpause (or what the exact rules are?)

grstepanov replied:
Specified what pause entails.

grstepanov committed:
Update LIP-3 based on new round of review

kadmil reviewed (approved):
LGTM

grstepanov committed:
Updated LIP-3 status to Approved
